Add mock module and build a mock bundle

btford opened the pull request:
I originally chatted with @tbosch about this – I'm not sure if there's a good way to test the bundled mock library. There may be other things I've missed.

Closes #2325

tbosch commented:
Looks good. However, I don't get the difference between `Location` and `BrowserLocation` / why we need 2 different classes. Couldn't we have an abstract base class that implements common behavior, and then specific ones for the browser / mock?

btford committed:
refactor(router): refactor BrowserLocation into LocationStrategy

This makes it easy to mock browser location and paves the way to
implementing hash routing.

btford committed:
feat(mock): add mock module and bundle

Closes #2325

tbosch commented on modules/angular2/src/mock/dummy_location_strategy.ts:
```diff
@@ -1,12 +1,9 @@
-@IMPLEMENTS(BrowserLocation)
-export class DummyBrowserLocation extends SpyObject {
+
+export class DummyLocationStrategy extends LocationStrategy {
```
I think this should be `MockLocationStrategy` instead of `Dummy...`

btford replied:
:+1:

tbosch commented on modules/angular2/src/router/location.ts:
```diff
@@ -1,44 +1,57 @@
   private _baseHref: string;
 
-  constructor(public _browserLocation: BrowserLocation,
+  constructor(public _platformLocation: LocationStrategy,
```
The fieldname should be `_platformStrategy`...

btford replied:
:+1:

btford commented:
Landed as e5de1f771a549e1bbe529890f5f8a3d61faac464 and 2932377769ca9365339a1edeb3a4d94bab5b8083.
